### PR TITLE
v3.1.0 seems to break webpack-dev-server behavior

### PR DESCRIPTION
**Pull request by ericclemmons**
**tl:dr** It appears that, in going from `3.0.0` to `3.1.0`, something has changed the way this package interacts with `webpack-dev-server` which causes the dev server to be unable to serve purely from memory. As a result, running webpack-dev-server on an app which has not already been transpiled once will not work.

For more details, see below:

For example, I have an existing app that uses `webpack-dev-server` and `npm-install-webpack-plugin@3.0.0`. Source files live in `src`, output lives in `dist`, the dev server uses `dist` as its contentBase, etc. If I delete the `dist` folder, and run the dev server, it will still serve up my files just fine, and respond to changes even though the `dist` folder doesn't actually exist. This simulates what (for example) a newly-scaffolded app might look like.

However, I created a new app last night that uses the exact same config and packages as my existing app, and suddenly `webpack-dev-server` would claim to be producing a valid bundle, but my browser would just throw 404 errors over and over. After troubleshooting the code itself, babel, my webpack config, restarting my computer, etc., I finally started diffing the `package.json` between the two apps and testing each difference by installing the older version of each package where the versions were different. Sure enough, rolling back to `npm-install-webpack-plugin@3.0.0` solves the problem entirely. Conversely, upgrading the older (and currently working) app the `3.1.0` will cause the webpack dev server problem to suddenly appear.

For reference, heres my `devDependendencies`:
```
  "devDependencies": {
    "babel-cli": "^6.7.5",
    "babel-eslint": "^6.0.0",
    "babel-loader": "^6.2.4",
    "babel-plugin-array-includes": "^2.0.3",
    "babel-plugin-transform-decorators-legacy": "^1.3.4",
    "babel-plugin-transform-object-assign": "^6.5.0",
    "babel-preset-es2015": "^6.6.0",
    "babel-preset-react": "^6.5.0",
    "babel-preset-react-hmre": "^1.1.1",
    "babel-preset-stage-1": "^6.5.0",
    "clean-webpack-plugin": "^0.1.8",
    "css-loader": "^0.23.1",
    "eslint": "^2.5.3",
    "eslint-config-airbnb": "^6.2.0",
    "eslint-plugin-react": "^4.2.3",
    "extract-text-webpack-plugin": "^1.0.1",
    "html-webpack-plugin": "^2.15.0",
    "mobx-react-devtools": "^4.0.2",
    "npm-install-webpack-plugin": "^3.0.0",
    "rimraf": "^2.5.2",
    "style-loader": "^0.13.1",
    "webpack": "^1.12.14",
    "webpack-dev-server": "^1.14.1",
    "webpack-merge": "^0.12.0"
  }
```

Let me know if you need any other info. Thanks!

### CODE REVIEW

**Comment by ericclemmons** [comment]
Ok, this is weird.  With this PR I created `example/webpack1-dev-server` and ran:

```shell
 ~/Projects/ericclemmons/npm-install-webpack-plugin/example/webpack1-dev-server  ⑂ 43-webpack-dev-server +    npm start

> @ prestart /Users/Eric/Projects/ericclemmons/npm-install-webpack-plugin/example/webpack1-dev-server
> rimraf build


> @ start /Users/Eric/Projects/ericclemmons/npm-install-webpack-plugin/example/webpack1-dev-server
> webpack-dev-server --config webpack.config.client.js --content-base=public --hot --inline --port=3000 --progress

Installing babel-core...
Installing babel-plugin-react-html-attrs...
Installing babel-preset-react...
Installing babel-preset-es2015...
Installing babel-preset-stage-0...
Installing babel-preset-react-hmre...
 70% 1/1 build moduleshttp://localhost:3000/
webpack result is served from /
content is served from /Users/Eric/Projects/ericclemmons/npm-install-webpack-plugin/example/webpack1-dev-server/public
Installing babel-loader...
 54% 6/8 build modulesInstalling bundle-loader...
 48% 45/71 build modulesInstalling style-loader...
Installing css-loader...
 68% 98/101 build modulesInstalling file-loader...
Installing url-loader...
Hash: dadf53f346c2025a1c82
Version: webpack 1.12.15
Time: 33779ms
                               Asset       Size  Chunks             Chunk Names
f4769f9bdb7466be65088239c12046d1.eot    20.1 kB          [emitted]
                           client.js     942 kB       0  [emitted]  client
                  1.dadf5-2a09093.js  711 bytes       1  [emitted]
chunk    {0} client.js (client) 882 kB [rendered]
    [0] multi client 52 bytes {0} [built]
    [1] (webpack)-dev-server/client?http://localhost:3000 2.67 kB {0} [built]
    ...
webpack: bundle is now VALID.
```

and got:
> screen shot 2016-04-29 at 12 48 56 pm

**It works**.

I'm going to have to clone the `react-boilerplate` project because there's definitely something different happening there...

**Comment by ericclemmons** [comment]
I'm starting to think this is something specific to `html-webpack-plugin` or `indexTemplate` part.

```shell
 ~/Projects/survivejs/react-boilerplate  ⑂ npm-install-fail +    npm start
# Original, no npm-install-webpack-plugin
> react-boilerplate@0.0.0 start /Users/Eric/Projects/survivejs/react-boilerplate
> webpack-dev-server

[webpack-validator] Config is valid.
http://localhost:8080/
webpack result is served from /
content is served from /Users/Eric/Projects/survivejs/react-boilerplate
404s will fallback to /index.html
webpack: wait until bundle finished: /
Child html-webpack-plugin for "index.html":

webpack: bundle is now VALID.
^C
 ~/Projects/survivejs/react-boilerplate  ⑂ npm-install-fail +    npm start
# With npm-install-webpack-plugin
> react-boilerplate@0.0.0 start /Users/Eric/Projects/survivejs/react-boilerplate
> webpack-dev-server

[webpack-validator] Config is valid.
http://localhost:8080/
webpack result is served from /
content is served from /Users/Eric/Projects/survivejs/react-boilerplate
404s will fallback to /index.html
webpack: wait until bundle finished: /
Child html-webpack-plugin for "index.html":

webpack: bundle is now VALID.
```

I get the 404 page, and when I add code like `import something from "something"`, I can tell the plugin is working, but the built `index.html` isn't being served up.

**<http://localhost:8080/app.js> and <http://localhost:8080/style.js> both still work!**

**Comment by bebraw** [comment]
@jantimon Any clue why html-webpack-plugin might fail here?

**Comment by ericclemmons** [comment]
Ok, I think I got it.  More coming...

**Comment by ericclemmons** [comment]
@bebraw Ok, the issue is with `html-webpack-plugin` caching the compilation result, manually adding it to the `compilation` result, which means that **the HTML generates during the dry-run, but isn't added to the _real_ compilation target**.

It's simply the plugin not using the correct hooks. Caching is fine, but it should _always_ add the output to the result.

18a7190 seems to solve it.

**Comment by ericclemmons** [comment]
@bebraw Dunno if you can apply this patch locally, but it worked for me:

```diff
diff --git a/.nvmrc b/.nvmrc
index 8470d3a..9773998 100644
--- a/.nvmrc
+++ b/.nvmrc
@@ -1 +1 @@
-v5.7.1
+v6.0.0
diff --git a/src/plugin.js b/src/plugin.js
index 3d5b7b2..5dd9522 100644
--- a/src/plugin.js
+++ b/src/plugin.js
@@ -55,11 +55,22 @@ NpmInstallPlugin.prototype.apply = function(compiler) {

 NpmInstallPlugin.prototype.preInstall = function(compilation, next) {
   var options = this.compiler.options;
-  var dryrun = webpack(Object.assign(
+  var config = Object.assign(
+    // Start with new config object
     {},
-    { cache: {} },
-    options
-  ));
+    // Inherit the current config
+    options,
+    {
+      // Ensure fresh cache
+      cache:{},
+      // Register plugin to install missing deps
+      plugins: [
+        new NpmInstallPlugin(this.options),
+      ],
+    }
+  );
+
+  var dryrun = webpack(config);

   dryrun.outputFileSystem = new MemoryFS();
```

**Comment by ericclemmons** [comment]
@cafreeman Are you running that plugin too?

**Comment by ericclemmons** [comment]
Unless someone says "this is broken!", I'm shipping soon :)

**Comment by cafreeman** [comment]
@ericclemmons Hi, sorry about the lag in responses. For some reason these notifications aren't showing up in my GitHub account :(

I am definitely running `html-webpack-plugin` as well.

**Comment by cafreeman** [comment]
@ericclemmons @bebraw ok, just installed from the `43-webpack-dev-server` branch, blew away my build folder and ran `npm start` successfully! Looks like it's fixed :)

**Comment by ericclemmons** [comment]
YES!! Thanks for the help guys!

I'll be releasing `3.1.1` today!

**Comment by jantimon** [comment]
Needs anything to be fixed in the html-webpack-plugin?

**Comment by ericclemmons** [comment]
@jantimon Nope! You can see the explanation & the fix here:
> https://github.com/ericclemmons/npm-install-webpack-plugin/pull/43#issuecomment-215860865

**Comment by ericclemmons** [comment]
`v3.1.1` is out!
